documentation/design/proposals/management_tasks-for-shared.adoc

```diff
@@ -0,0 +1,91 @@
+1. The Console UI uses modals UI elements to initiate the request and blocks awaiting the response. The user is unable
+   to continue to interact with the app whilst the request is in progress.
+1. There is no provision for a command line interface.
+1. Poorly defined RBAC model.
+
+== High Level Proposal
+
+Introduce a new CRD `ManagementTask`.  A `ManagementTask` executes a management function (such as a purge) in a
+tenant namespace within EnMasse.  It would be rather similar to
+https://kubernetes.io/docs/concepts/workloads/controllers/jobs-run-to-completion/[Jobs] from Kubernetes.
+
+1. The Tenant Admin would create an instance of the resource in order to cause the task to occur.  The contents of the
+   resource describe the jobs and its parameters.
+1. The EnMasse Operator actions the tasks - initiating requests to Brokers or Routers as necessary.  This might invoke
+  the invocation of Router/Broker management functions (close connection, clear queue etc), or might involve putting down
+  router wiring to implement a move message task.
+1. The EnMasse Operator periodically updates the task with progress of the task thus allowing the Tenant Admin to monitor
+   the task for completion.  The availablility of progress information depends on the type of task.
+1. The Tenant Admin may delete a ManagementTask, this would be a signal to the Operator to cancel the task (if possible).
+1. The Operator would mark the ManagementTask as compete. A `ttlSecondsAfterFinished` parameter (defined at `MessagingInfra`
+   level) would allow the Operator to clean up the ManagementTasks automatically.
+
+The tenant admin would be able to CRUDWL `ManagementTask` from the CLI using `kubectl`/`oc` tooling (usual RBAC permissions
+apply), or later an EnMasse CLI tool might abstract away some details.
+
+From the Console. functions such as purge queue/subscription would, behind the button cause an instance of this object to
+be created.  There would be an area of the UI which lists running/recently finsihed tasks.  From this page, you'd be able
+to cancel running tasks too.
+
+=== CRD sketch
+
+```
+apiVersion: enmasse.io/v1
+kind: ManagementTask
+metadata:
+  name: mytenantnamespace
+spec:
+  # oneOf
+  addressPurge:
+    addressSelector:
+      matchLabels:
+        foo: bar
+      matchNames:
+       - addr1
+       - addr1
+  connectionClose:  
+    connectionSelector:
+      matchNames:
+       - example.com:3456
+       - example.com:4567
+  moveMessage:
+     sourceAddress: addr1
+     filter:
+       type: amqp:sql-filter
+       expression: application-properties.color = 'blue'
+     targetAddress: add2
+status:
+  phase: New | Running | Failed | Completed
+  message: ..
+  conditions:
+    ...
+```
+
+Open Questions:
+
+* The ability of create a ManagementTask would be typically granted to an Tenant Admin.  This would give that user
+  the permission to do any type of management functiion permitted by the task.  Would we need to be more discriminating?
```

Comment:
We could add another 'verb' to the MessagingRole called 'manage', which, when granted would allow you to perform purge and move operations on addresses where you have that role, which is similar to what we have today. That would allow creating more fine-grained RBAC rules. However, I don't think that would change the above design fundamentally.

Reply:
I was seeing this as being permissioned by the Kubernetes domain rather than the EnMasse Messaging permissions model (MessagingRole/MessagingRoleBinding).   What is your idea?

I was wondering in [RBAC subresources](https://kubernetes.io/docs/reference/access-authn-authz/rbac/#referring-to-resources) could help.  In order to be able to purge an address, the EnMasse Operator would check for a permission granting `address/purge` on the target resource.  My difficulty is I don't see a way to find out who created the ManagementTask.

Reply:
@k-wall Yeah, I was thinking to use the messaging permissions model, but maybe it would complicate things too much. It could be that RBAC on ManagementTask is sufficient given that a namespace only have 1 messaging tenant anyway. I'm not too familiar of who is typically given these permissions in an organization.

It would be nice to use subresources for this, I'm not sure if it would be allowed to create generic CRD subresources that are not `scale` or `status`.

Reply:
@lulf My gut feel is that RBAC would be sufficient.  In other words, I think as tenant admin should be able to create any type of ManagementTask.